src/main/java/duke/Duke.java

```diff
@@ -0,0 +1,68 @@
+package duke;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.PrintWriter;
+
+public class Duke {
+    private String taskListFileName;
+
+    /**
+     * Configures a Duke runtime to use a taskList storage file at specified location.
+     * @param taskListFileName path to taskList storage file.
+     */
+    public Duke(String taskListFileName) {
+        this.taskListFileName = taskListFileName;
+    }
+
+    public static class DukeException extends Exception {
+        public DukeException(String errorMessage) {
+            super(errorMessage);
+        }
+    }
+
+    /**
+     * Default way to run the Duke programme.
+     */
+    public void run() {
+        this.run(new BufferedReader(new InputStreamReader(System.in)), new PrintWriter(System.out));
+    }
+
+    /**
+     * Runs the Duke programme with provided inputs and outputs. Normally invoked by run() unless dependencies need
+     * to be injected by test.
+     * @param in BufferedReader taking in user input.
+     * @param out PrintWriter to write responses to user input.
+     */
+    public void run(BufferedReader in, PrintWriter out) {
+        Ui ui = new Ui(in, out);
+        Parser parser = new Parser(ui);
+        Storage storage = new Storage(this.taskListFileName);
+        TaskList taskList = storage.initialise();
+
+        ui.init();
+        boolean ended = false;
```

Comment:
Should a prefix be added to this variable name to make it sound more like a boolean? One example is: 
```suggestion
        boolean hasEnded = false;
```

Reply:
Amended as of 4030096e42b9fe4bd5c9ccecfed49e72b055c620

Reply:
amended as of 734dcb9b1359b25907ac79f9f742e0c796f93a03